python/tutorials/gluon/09-tma-gather-scatter.py

```diff
@@ -0,0 +1,658 @@
+    out_desc_layout = gl.NVMMASharedLayout.get_default_for([BLOCK_M, BLOCK_N], dtype)
+
+    X_desc = TensorDescriptor.from_tensor(X, [1, BLOCK_K], X_desc_layout)
+    W_desc = TensorDescriptor.from_tensor(W, [BLOCK_K, BLOCK_N], W_desc_layout)
+    out_desc = TensorDescriptor.from_tensor(out, [1, BLOCK_N], out_desc_layout)
+
+    # Persistent kernel grid.
+    num_sms = torch.cuda.get_device_properties("cuda").multi_processor_count
+    num_pid = triton.cdiv(M, BLOCK_M) * triton.cdiv(N, BLOCK_N)
+    grid = (min(num_sms, num_pid), )
+    SchedulerImpl = t7.GroupedPersistentTileScheduler(GROUP_SIZE_M)
+    matmul_fused_gather_scatter_kernel[grid](X_desc, W_desc, out_desc, X_gather_indx, out_scatter_indx, BLOCK_M,
+                                             SchedulerImpl, num_buffers)
+    return out
+
+
+@pytest.mark.parametrize("M, N, K", [(1024, 1024, 2048), (4096, 4096, 4096)])
+@pytest.mark.parametrize("BLOCK_M, BLOCK_N", [(128, 128), (128, 64)])
+@pytest.mark.parametrize("BLOCK_K, num_buffers", [(128, 2), (64, 3)])
+@pytest.mark.skipif(not is_blackwell(), reason="Requires Blackwell")
+def test_matmul_fused_gather_scatter(M, N, K, BLOCK_M, BLOCK_N, BLOCK_K, num_buffers):
+    torch.manual_seed(0)
+
+    # Randomize the gather indices.
+    X_gather_indx = torch.arange(0, M, dtype=torch.int32, device="cuda")
+    shfl = torch.randperm(M, device="cuda")
+    X_gather_indx = X_gather_indx[shfl]
+
+    # Randomize the scatter indices.
+    out_scatter_indx = torch.arange(0, M, dtype=torch.int32, device="cuda")
+    shfl = torch.randperm(M, device="cuda")
+    out_scatter_indx = out_scatter_indx[shfl]
+
+    X = torch.randn(M, K, dtype=torch.bfloat16, device="cuda")
+    W = torch.randn(K, N, dtype=torch.bfloat16, device="cuda")
+    out = matmul_fused_gather_scatter(X, X_gather_indx, W, out_scatter_indx, BLOCK_M, BLOCK_N, BLOCK_K,
+                                      num_buffers=num_buffers)
+
+    out_ref = torch.empty_like(out)
+    out_ref[out_scatter_indx, :] = X[X_gather_indx, :] @ W
+    torch.testing.assert_close(out, out_ref, atol=1e-3, rtol=1e-3)
+
+
+# %%
+# The main takeaway from this tutorial is understanding how to use `async_gather`
+# and `async_scatter`. These instructions provide a middle-ground between
+# block DMAs like `async_copy_global_to_shared` and `async_copy_shared_to_global`
+# and regular global loads and stores (`gl.load` and `gl.store`) by allowing
+# separately-indexed columns while maintaining the performance of TMAs.
+#
+# Keep in mind the following:
+# - `async_gather` and `async_scatter` are typically faster than `gl.load` and
+#   `gl.store` when they can be used, but this is not always the case. Plus, TMA
+#   instructions use shared memory.
+# - Sometimes using `async_gather` or `async_scatter` instead of block DMA
+#   instructions like `async_copy_global_to_shared` and `async_copy_shared_to_global`
+#   is actually faster, but these situations are rare.
```

Comment:
As in `gather(x_offset + arange(0, X_BLOCK), y_offset)` is faster than `load([x_offset, y_offset])`? If so, that's bizarre.

Reply:
Yes. We discovered this when first integrating fused gather into matmul_ogs. It depends on the block size and has something to due with the TMA messages and L2 cache but we never got to the bottom of it. I just want to call it out because while counterintuitive, it's another performance knob to tune.